### ISSUE TITLE
Array.prototype.shuffle

### ISSUE THREAD

**Issue by funforums** [nanodeath/CrowLib#18]
Is it really necessary to have that tweak in the Array.prototype property?
It has the mean consequence that every for...in loop in the code of anybody using this library will fail if not checking for hasOwnProperty in EVERY loop body, making the code look way uglier in big projects.

Moreover, i thought extending native types prototypes was generally bad practice, probably for this reason.

**Comment by unstoppablecarl** [comment]
It is bad practice.

**Comment by nanodeath** [comment]
Sure, I agree.  I also believe it's a best practice to call `hasOwnProperty` when iterating over keys of an object, though, verbose or not.

**Comment by funforums** [comment]
That's true, it probably is. Though while hasOwnProperty can be used as a best practice, it will not certainly break the code if not used, whereas extending native types' prototypes has a very high chance of breaking backwards-compatibility :)

It also has the risk that anybody getting this library to integrate it with their project may discard it as soon as they notice it breaks their code, which is a pity, because I honestly find it very good.

**Comment by nanodeath** [comment]
@funforums Would you be open to submitting a pull request, or should I look into this when I have time?

**Comment by funforums** [comment]
I might help out, if you want, but since it is your project, I would first agree on a fix with you :)

Is it ok if I propose a fix and then, if you agree, I can apply it?

**Comment by nanodeath** [comment]
That's absolutely fine.  Don't get me wrong, I totally agree that modifying core prototypes is evil and wouldn't do it again if I were to write this library from scratch now.  But I also feel if one programs defensively, this wouldn't be a problem.  I can understand if not everyone wants to code that way, though.

**Comment by unstoppablecarl** [comment]
It is not reasonable to expect devs to code defensively to accommodate your code. Development is hard enough without having to _defend_ myself from dependency code by wrapping it.

``` js
 if(doesNotConformToReasonableExpectation) { 
    /* @TODO insert workaround */
 }
```

**Comment by nanodeath** [comment]
My opinion is always that should _always_ use hasOwnProperty for `for(... in ...)` loops because you never know what libraries you're using are going to do, but I don't think that's relevant.  I already said I think we should make the change.

**Comment by funforums** [comment]
@nanodeath I have opened a pull request.

**Comment by nanodeath** [comment]
Was resolved ages ago in #19.
